TLS Support for Bacula

RaHoni opened the pull request:
## Add support for TLS in the Bacula module

This PR adds support to configure TLS in Bacula. 

Sorry for the mass ping in #245923 

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

RaHoni commented:
Pinging @therealr5  @JulianFP since they contributed  before the mass ping

JulianFP reviewed (approved):
diff confirms that the changes in the files are exactly the same, so I still approve

quantenzitrone commented on nixos/modules/services/backup/bacula.nix:
```diff
@@ -108,13 +128,95 @@ let
+        type = types.nullOr types.bool;
+        default = null;
+        description = mdDoc ''
+          Require TLS or TLS-PSK encryption. This directive is ignored unless
+          one of TLS Enable or TLS PSK Enable is set to true.
+          If TLS is not required while TLS or TLS-PSK are enabled, then the
+          Bacula component will connect with other components either with or
+          without TLS or TLS-PSK
+
+          If TLS or TLS-PSK is enabled and TLS is required, then the Bacula
+          component will refuse any connection request that does not use TLS.
```
I think this description is a bit confusing. 
TLS enable is never set to `true`, as the Bacula configuration syntax uses yes/no. `${submodulePath}.tls.enable` is set to `true`. 
TLS-PSK enable is not exposed as a NixOS option, so how is a user supposed to set that to true?

quantenzitrone commented on nixos/modules/services/backup/bacula.nix:
```diff
@@ -108,13 +128,95 @@ let
+          component will refuse any connection request that does not use TLS.
+        '';
+      };
+      certificate = mkOption {
+        type = types.nullOr types.str;
+        default = null;
+        description = mdDoc ''
+          The full path and filename of a PEM encoded TLS certificate.
+          It will be used as either a client or server certificate,
+          depending on the connection direction.
+          This directive is required in a server context, but it may
+          not be specified in a client context if TLS Verify Peer is
+          set to `false` in the corresponding server context.
+        '';
+      };
+      key = mkOption {
+        type = types.str;
+        description = mdDoc ''
+          The full path and filename of a PEM encoded TLS
+          private key. It must correspond to the TLS certificate.
+        '';
+      };
```
I think "path" describes "full path and filename" enough.
You can also probably use `types.path` instead of `types.str`.

quantenzitrone commented on nixos/modules/services/backup/bacula.nix:
```diff
@@ -9,6 +9,21 @@ with lib;
 let
   libDir = "/var/lib/bacula";
 
```
```suggestion
  inherit (lib)
    concatStringsSep
    literalExpression
    mdDoc
    mkOption
    optional
    optionalString
    types
    # and any other used functions from lib
    ;
```

+ remove the `with lib;`

https://nix.dev/guides/best-practices#with-scopes

quantenzitrone commented on nixos/modules/services/backup/bacula.nix:
```diff
@@ -108,13 +128,95 @@ let
+        '';
+      };
+      caCertificateFile = mkOption {
+        type = types.nullOr types.str;
```
maybe
```suggestion
        type = types.nullOr types.path;
```

quantenzitrone reviewed (changes requested):
The descriptions of a lot of the TLS options reference direct Bacula config attributes, which requires the user to look up the module source code to figure out what is going on.
Since the TLS submodule is used in 2 modules, i would introduce a function argument `submodulePath` or something, that then is used in the option descriptions to directly reference the relevant NixOS options instead of Bacula config attributes.
Something like:
```nix
tlsOptions = submodulePath: {...}: {
  someOption = mkOption {
    description = ''
      Will have no effect if `${submodulePath}.someOtherOption` is set to `true`.
    '';
  };
}:
```

I asked on matrix about multiline module descriptions, and while there is no official convention, it is recommended to use at most [one sentence per line](https://nix.dev/contributing/documentation/style-guide#one-line-per-sentence), and semantic line feeds on long sentences. This is what the documentation team uses. So maybe reformat that accordingly.

RaHoni commented:
> I asked on matrix about multiline module descriptions, and while there is no official convention, it is recommended to use at most [one sentence per line](https://nix.dev/contributing/documentation/style-guide#one-line-per-sentence), and semantic line feeds on long sentences. This is what the documentation team uses. So maybe reformat that accordingly.

Just asking if I should fix the formatting also in the existing documentation or only in the new?

Also should I run `nixpkgs-fmt` on evrything?

quantenzitrone commented:
I'd say just the new, and no, don't run `nixpkgs-fmt`. Nixpkgs does not yet have an official formatter. There currently is https://github.com/nixos/rfcs/pull/166 trying to establish an official format also creating a formatter implementation, but that can take another while.

RaHoni committed:
nixos/bacula: Add support for TLS

Added the majority of TLS options used for transport encryption.

This uses a submodule since all resources share the same settings.

The documentation can be found under:
https://www.bacula.org/13.0.x-manuals/en/main/Bacula_TLS_Communications_E.html

RaHoni committed:
nixos/bacula: refactor option generation

The generation of the descriptions always used lib explicitly even
thoughit was already specified with 'with'.

Since using `with lib` is discouraged this was also changed to inherit.

RaHoni commented:
Can I keep the commits like they are right now? Because otherwise I would risk a new mass ping.

quantenzitrone commented on nixos/modules/services/backup/bacula.nix:
```diff
@@ -108,13 +138,98 @@ let
+          Additionally, the client's X509 certificate Common Name must meet the value of the Address directive.
+          If the TLSAllowed CN configuration directive is used,
+          the client's x509 certificate Common Name must also correspond to
+          one of the CN specified in the TLS Allowed CN directive.
```
```suggestion
          one of the CN specified in the ${tlsLink "tls.allowedCN" submodulePath} directive.
```

quantenzitrone commented on nixos/modules/services/backup/bacula.nix:
```diff
@@ -108,13 +138,98 @@ let
+          It will be used as either a client or server certificate,
+          depending on the connection direction.
+          This directive is required in a server context, but it may
+          not be specified in a client context if TLS Verify Peer is
```
```suggestion
          not be specified in a client context if ${tlsLink "tls.verifyPeer" submodulePath} is
```

quantenzitrone commented on nixos/modules/services/backup/bacula.nix:
```diff
@@ -108,13 +138,98 @@ let
+          This directive is valid for a server and in a client context.
+          If this directive is specified, the peer certificate will be verified against this list.
+          In the case this directive is configured on a server side, the allowed
+          CN list will not be checked if TLS Verify Peer is set to false.
```
...

quantenzitrone commented on nixos/modules/services/backup/bacula.nix:
```diff
@@ -108,13 +138,98 @@ let
+        description = mdDoc ''
+        The path specifying a PEM encoded TLS CA certificate(s).
+        Multiple certificates are permitted in the file.
+        One of TLS CA Certificate File or TLS CA Certificate Dir are required in a server context, unless TLS
+        verifyPeer is set to no, and are always required in a client context.
```
```suggestion
        One of TLS CA Certificate File or TLS CA Certificate Dir are required in a server context, 
        unless ${tlsLink "tls.verifyPeer" submodulePath} is set to false, and are always required in a client context.
```

quantenzitrone commented on nixos/modules/services/backup/bacula.nix:
```diff
@@ -324,16 +446,26 @@ in {
+        default = { };
+        description = mdDoc ''
+          TLS Options for the File Daemon.
+          Important notice the Backup won't get encrypted.
```
```suggestion
          Important notice: The Backup won't get encrypted.
```

quantenzitrone commented on nixos/modules/services/backup/bacula.nix:
```diff
@@ -496,13 +636,22 @@ in {
+        default = { };
+        description = mdDoc ''
+          TLS Options for the Director.
+          Important notice the Backup won't get encrypted.
```
```suggestion
          Important notice: The Backup won't get encrypted.
```

quantenzitrone reviewed (changes requested):
Not sure if force pushing a commit with the same commit message as a previous commit triggers a new mass ping. I'd think not, since this is a rather common practice in NixPkgs. Maybe wait for someone with merge privileges, they are the most knowledgeable about this stuff.

quantenzitrone commented on nixos/modules/services/backup/bacula.nix:
```diff
@@ -108,13 +138,99 @@ let
+          This directive is valid for a server and in a client context.
+          If this directive is specified, the peer certificate will be verified against this list.
+          In the case this directive is configured on a server side, the allowed
+          CN list will not be checked if ${tlsLink "tls.verifyPeer"} is false.
```
```suggestion
          CN list will not be checked if ${tlsLink "tls.verifyPeer" submodulePath} is false.
```

manual fails to build because of this

RaHoni replied:
Thanks for finding it otherwise I would have searched it my self?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1351

rouven0 reviewed (approved):
LGTM, have you ever tested a bacula config using your new tls options?

JulianFP reviewed (approved):
Tested generation of config files in VM and and I built and inspected the manual just to make sure. Everything still looks good to me.

RaHoni commented:
> LGTM, have you ever tested a bacula config using your new tls options?

As you can see at https://github.com/RaHoni/nixos-conf I use it for my Desktop in Production.

quantenzitrone commented:
you gave up?

RaHoni commented:
No I actualy didn't notice I simply renamed the branch since I want'ed to do another PR and this being master was kinda unfortuned.
